File: packages/guards/src/semantic-coherence.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface SemanticCoherenceOptions {
  action: 'block' | 'warn';
}

const CONTRADICTION_PAIRS: [RegExp, RegExp][] = [
  [/\bis\s+true\b/i, /\bis\s+false\b/i],
  [/\balways\b/i, /\bnever\b/i],
  [/\byes\b/i, /\bno\b/i],
  [/\bincreased?\b/i, /\bdecreased?\b/i],
];

function getSentences(text: string): string[] {
  return text.split(/[.!?]+/).map(s => s.trim()).filter(s => s.length > 3);
}

function wordSaladScore(text: string): number {
  const words = text.toLowerCase().split(/\s+/);
  if (words.length < 5) return 0;
  const unique = new Set(words);
  const ratio = unique.size / words.length;
  if (ratio > 0.95 && words.length > 20) return 0.8;
  return 0;
}

export function semanticCoherence(options: SemanticCoherenceOptions): Guard {
  return {
    name: 'semantic-coherence',
    version: '0.1.0',
    description: 'Basic semantic coherence check for responses',
    category: 'ai',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];
      const sentences = getSentences(text);

      for (let i = 0; i < sentences.length - 1; i++) {
        const a = sentences[i];
        const b = sentences[i + 1];
        for (const [p1, p2] of CONTRADICTION_PAIRS) {
          if ((p1.test(a) && p2.test(b)) || (p2.test(a) && p1.test(b))) {
            issues.push('contradiction');
            break;
          }
        }
      }

      const salad = wordSaladScore(text);
      if (salad > 0.5) issues.push('word-salad');

      const triggered = issues.length > 0;
      const score = triggered ? Math.min((issues.length * 0.4) + salad, 1.0) : 0;

      return {
        guardName: 'semantic-coherence',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/semantic_coherence.py
"""Basic semantic coherence check for responses."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_CONTRADICTION_PAIRS: list[
    tuple[re.Pattern[str], re.Pattern[str]]
] = [
    (
        re.compile(r"\bis\s+true\b", re.IGNORECASE),
        re.compile(r"\bis\s+false\b", re.IGNORECASE),
    ),
    (
        re.compile(r"\balways\b", re.IGNORECASE),
        re.compile(r"\bnever\b", re.IGNORECASE),
    ),
    (
        re.compile(r"\byes\b", re.IGNORECASE),
        re.compile(r"\bno\b", re.IGNORECASE),
    ),
    (
        re.compile(
            r"\bincreased?\b", re.IGNORECASE
        ),
        re.compile(
            r"\bdecreased?\b", re.IGNORECASE
        ),
    ),
]


def _get_sentences(text: str) -> list[str]:
    return [
        s.strip()
        for s in re.split(r"[.!?]+", text)
        if len(s.strip()) > 3
    ]


def _word_salad_score(text: str) -> float:
    words = text.lower().split()
    if len(words) < 5:
        return 0.0
    unique = set(words)
    ratio = len(unique) / len(words)
    if ratio > 0.95 and len(words) > 20:
        return 0.8
    return 0.0


class _SemanticCoherence:
    def __init__(
        self, *, action: str = "warn"
    ) -> None:
        self.name = "semantic-coherence"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []
        sentences = _get_sentences(text)

        for i in range(len(sentences) - 1):
            a, b = sentences[i], sentences[i + 1]
            for p1, p2 in _CONTRADICTION_PAIRS:
                if (
                    p1.search(a) and p2.search(b)
                ) or (
                    p2.search(a) and p1.search(b)
                ):
                    issues.append("contradiction")
                    break

        salad = _word_salad_score(text)
        if salad > 0.5:
            issues.append("word-salad")

        triggered = len(issues) > 0
        score = (
            min(len(issues) * 0.4 + salad, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (
            time.perf_counter() - start
        ) * 1000

        return GuardResult(
            guard_name="semantic-coherence",
            passed=not triggered,
            action=(
                self.action if triggered else "allow"
            ),
            score=score,
            message=(
                "Semantic coherence issue detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"issues": issues}
                if triggered
                else None
            ),
        )


def semantic_coherence(
    *, action: str = "warn"
) -> _SemanticCoherence:
    return _SemanticCoherence(action=action)
